File: 01-Programming-Languages/Ruby/Data-Structures/Stack/stack.rb
class Stack
  def initialize
    @items = []
  end

  def push(item)
    @items.push(item)
  end

  def pop
    @items.pop
  end

  def peek
    @items.last
  end

  def empty?
    @items.empty?
  end

  def size
    @items.length
  end
end

stack = Stack.new
stack.push(1)
stack.push(2)
stack.push(3)

puts "Stack size: #{stack.size}"
puts "Popped: #{stack.pop}"
puts "Top: #{stack.peek}"

File: 01-Programming-Languages/Scala/Data-Structures/Stack/stack.scala
class Stack[T] {
  private var items: List[T] = List()

  def push(item: T): Unit = {
    items = item :: items
  }

  def pop(): Option[T] = {
    items match {
      case Nil => None
      case head :: tail =>
        items = tail
        Some(head)
    }
  }

  def peek(): Option[T] = items.headOption

  def isEmpty: Boolean = items.isEmpty

  def size: Int = items.length
}

object StackApp extends App {
  val stack = new Stack[Int]()
  stack.push(1)
  stack.push(2)
  stack.push(3)

  println(s"Stack size: ${stack.size}")
  println(s"Popped: ${stack.pop().getOrElse("empty")}")
  println(s"Top: ${stack.peek().getOrElse("empty")}")
}